src/connection-loop.ts

```diff
@@ -0,0 +1,229 @@
+import {resolver} from './resolver.js';
+import {sleep} from './sleep.js';
+
+const DEBOUNCE_DELAY = 10;
+
+const MIN_DELAY = 30;
+const MAX_DELAY = 60_000;
+
+const MAX_CONNECTIONS = 3;
+
+type Duration = {duration: number; ok: boolean};
+type Durations = Duration[];
+
+export interface ConnectionLoopDelegate {
+  closed(): boolean;
+  invokeExecute(): unknown;
+  debounceDelay?: number;
+  maxConnections?: number;
+  printTime(): string;
+  watchdogTimer?: number;
+}
+
+export class ConnectionLoop {
```

Comment:
I think this is pretty straightforward and easy to understand, but in order to do it, I had to read through it and draw a state diagram in a notebook. You could help the reader out, by doing that for them:

```
ConnectionLoop runs a loop sending network requests (either pushes or pulls) to
the server. Our goal, generally, is to send requests as fast as we can, but to adjust in case of slowness, network errors, etc. We will send requests in parallel if the server
supports it. We also debounce pushes since they frequently happen in series very 
near to one another (e.g., during drag drops).

The loop flows through the following states forever, until it is closed:

Pending: Wait for event or watchdog
          |
          v
Debounce: Wait for more events (we debounce pushes)
          |
          v
Wait for avail connect (we limit number of parallel requests allowed)
          |
          v
Wait to send (if requests are taking too long, we will slow down)
          |
          v
Send (asynchronously, wrt the loop)
          |
          v
Back to the pending!
```

Reply:
Thanks